src/dune_rules/codept.ml

```diff
@@ -0,0 +1,105 @@
+  let dep = Obj_dir.Module.dep obj_dir in
+  let context = Super_context.context sctx in
+  let immediate_file = dep (Immediate source) in
+  let m2l_file = dep (M2l (unit, ml_kind)) in
+  let approx_dep_file = dep (Immediate_approx source) in
+  let sig_file m = dep (Sig m) in
+  let gen_sig =
+    match ml_kind with
+    | Intf -> true
+    | Impl -> not (Module.has unit ~ml_kind:Intf)
+  in
+  let flags, sandbox =
+    Option.value (Module.pp_flags unit)
+      ~default:(Action_builder.return [], sandbox)
+  in
+  let open Memo.O in
+  let* codept = codept_prog ~dir sctx in
+  let* () =
+    (* 1. Generate m2l and approx immediate from source. *)
+    Super_context.add_rule sctx ~dir
+      (let open Action_builder.With_targets.O in
+      Command.run codept
+        ~dir:(Path.build context.build_dir)
+        [ As [ "-k"; "-verbosity"; "error" ]
+        ; Command.Args.dyn flags
+        ; Command.Ml_kind.flag ml_kind
+        ; Dep (Module.File.path source)
+        ; S (codept_o_arg "-m2l" m2l_file)
+        ; S (codept_o_arg "-nl-modules" approx_dep_file)
+        ]
+      >>| Action.Full.add_sandbox sandbox)
+  in
+  let* () =
+    (* 2. Generate immediate and sig from m2l and approx immediate sigs. *)
+    let { Modules.vlib = vlib_modules; _ } = Modules.split_by_lib modules in
+    let is_vlib_module m =
+      (* TODO: better way to check this? or could include vlib modules? *)
+      List.exists vlib_modules ~f:(fun vm ->
+          Module_name.Unique.compare (Module.obj_name vm) (Module.obj_name m)
+          = Eq)
+    in
+    let build_paths dependencies =
+      let dependency_file_path m =
+        if Module.kind m = Alias || is_vlib_module m then None
+        else Some (Path.build (sig_file m))
+      in
+      List.filter_map dependencies ~f:dependency_file_path
+    in
+    let action =
+      let open Action_builder.O in
+      let paths =
+        let+ lines = Action_builder.lines_of (Path.build approx_dep_file) in
+        let modules = Dep_gen.interpret_deps md ~unit lines in
+        build_paths modules
+      in
+      let path_args =
+        let+ paths = paths in
+        Command.Args.Deps paths
+      in
+      let sig_args : _ Command.Args.t list =
+        if gen_sig then codept_o_arg "-sig" (sig_file unit) else []
+      in
+      let open Action_builder.With_targets.O in
+      Command.run codept
+        ~dir:(Path.build context.build_dir)
+        [ As [ "-k"; "-verbosity"; "error" ]
+          (* avoid self-cycle errors and unresolved module notifications *)
+        ; Command.Args.dyn flags
+        ; Dep (Path.build m2l_file)
+        ; Dyn path_args
+        ; S sig_args
+        ; S (codept_o_arg "-modules" immediate_file)
+        ]
+      >>| Action.Full.add_sandbox sandbox
+    in
+    Super_context.add_rule sctx ~dir action
+  in
+  let+ () =
+    (* 3. Merge transitives. *)
+    let file = Path.build m2l_file in
+    Dep_gen.transitive_of_immediate_rule md ~ml_kind ~source ~file unit
```

Comment:
Do we still need this with codept? 🤔 I thought codept returned the precise set of deps.

Reply:
Oh, I don't actually know. I just swapped out the pipeline up to the immediate `.d` files, not really knowing why the transitives are combined at the file level.

Codept has some `-expand-deps` option, which maybe would achieve the same? I also don't really know what exact/precise is supposed to mean in this context because transitive closure sounds like a less precise thing to me :sweat_smile:.

Reply:
This is the relevant comment: https://github.com/ocaml/dune/pull/666#issuecomment-377592847

I guess we should confirm with @Octachron

Reply:
Codept doesn't return the transitive closure of dependencies currently since there are some values for humans to have a minimal set of dependencies. This would be straightforward to add.

The `-expand-deps` option is an untested intermediary solution that count as immediate dependencies all type-level dependencies. For instance,  with `-expand-deps`
```ocaml
(* a.ml *)
type 'a t  = A of 'a
```
```ocaml
(* b.ml *)
type 'a u = 'a A.t = A of 'a 
val x: int u
```
```ocaml
(* c.ml *)
let y = B.x
```
the module `A` is considered to be an immediate dependency of `C`.

Reply:
Indeed this is still required for the test of #666 to pass. It's easiest to leave it as is for consistency, the transitive logic is still needed for `ocamldep` and this just reuses it.